Safety/Src/imu.cpp

```diff
@@ -1,119 +1,256 @@
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, MAG_IF_3_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) REP_XY_REGULAR_PRESET, 1, HAL_MAX_DELAY);
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, MAG_IF_2_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) MAG_REPXY_REG, 1, HAL_MAX_DELAY);
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, MAG_IF_3_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*)REP_Z_REGULAR_PRESET, 1, HAL_MAX_DELAY);
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, MAG_IF_2_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) MAG_REPZ_REG, 1, HAL_MAX_DELAY);
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, MAG_IF_3_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) MAG_IF_3_DATA_MODE, 1, HAL_MAX_DELAY);
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, MAG_IF_2_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) MAG_IF_2_DATA_MODE, 1, HAL_MAX_DELAY);
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, MAG_IF_1_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) MAG_IF_1_DATA_MODE, 1, HAL_MAX_DELAY);
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, MAG_CONF_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) MAG_REFRESH_200_HZ, 1, HAL_MAX_DELAY);
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, CMD_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) MAG_SETUP_DIS, 1, HAL_MAX_DELAY);
 
 }
 
+bool scan() {
+	if (HAL_I2C_IsDeviceReady(&hi2c1, BMX160_I2C_ADDR, 50, HAL_MAX_DELAY) == HAL_OK) {
+		return true;
+	}
+	return false;
+}
 
-/************************ (C) COPYRIGHT STMicroelectronics *****END OF FILE****/
+void BMX160::calibrate(void) {
+	const int nSamplesForReliableAverage = 100;
+    IMUData_t TempImuData;
+
+    IMUCalibration.accel_x = 0.0f;
+    IMUCalibration.accel_y = 0.0f;
+    IMUCalibration.accel_z = 0.0f;
+    IMUCalibration.gyro_x = 0.0f;
+    IMUCalibration.gyro_y = 0.0f;
+    IMUCalibration.gyro_z = 0.0f;
+    IMUCalibration.mag_x = 0.0f;
+    IMUCalibration.mag_y = 0.0f;
+    IMUCalibration.mag_z = 0.0f;
+
+    for (int i = 0; i < nSamplesForReliableAverage; i++)
+    {
+        this->updateData();
+
+        HAL_Delay(7);
+
+        this->GetResult(TempImuData);
+
+        IMUCalibration.gyro_x += TempImuData.gyro_x;
+        IMUCalibration.gyro_y += TempImuData.gyro_y;
+        IMUCalibration.gyro_z += TempImuData.gyro_z;
+        IMUCalibration.accel_x += TempImuData.accel_x;
+        IMUCalibration.accel_y += TempImuData.accel_y;
+        IMUCalibration.accel_z += TempImuData.accel_z;
+
+    }
+
+    IMUCalibration.gyro_x /= (float) nSamplesForReliableAverage;
+    IMUCalibration.gyro_y /= (float) nSamplesForReliableAverage;
+    IMUCalibration.gyro_z /= (float) nSamplesForReliableAverage;
+    IMUCalibration.accel_x /= (float) nSamplesForReliableAverage;
+    IMUCalibration.accel_y /= (float) nSamplesForReliableAverage;
+    IMUCalibration.accel_z /= (float) nSamplesForReliableAverage;
+
+    IMUCalibration.accel_z -= 1000.0f;    // at calibration, Z needs to read -1g.
```

Comment:
shouldn't z be at -1 already (since the tempimudata should return the resolved value)?